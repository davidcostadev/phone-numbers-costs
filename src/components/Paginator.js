import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { paginatorType } from '../types';

const Item = ({
  id,
  page,
  text,
  onClickPage,
  isCurrent,
  isDisabled,
}) => (
  <button
    type="button"
    id={`paginator-${id}`}
    className={classnames(
      'paginator__item',
      { 'paginator__item--active': isCurrent },
      { 'paginator__item--disabled': isDisabled },
    )}
    disabled={isDisabled}
    onClick={() => onClickPage(page)}
  >
    {text}
  </button>
);

Item.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  page: PropTypes.number,
  text: PropTypes.string.isRequired,
  onClickPage: PropTypes.func.isRequired,
  isCurrent: PropTypes.bool,
  isDisabled: PropTypes.bool,
};

Item.defaultProps = {
  isCurrent: false,
  isDisabled: false,
  page: null,
};


const List = ({ currentPage, pagesList, onClickPage }) => pagesList
  .map(page => (
    <Item
      key={page}
      id={page}
      page={page}
      text={page.toString()}
      isCurrent={page === currentPage}
      onClickPage={onClickPage}
    />
  ));

const Previous = ({ previousPageExists, previousPage, onClickPage }) => (
  <Item
    id="previous"
    page={previousPage}
    text="Previous"
    isDisabled={!previousPageExists}
    onClickPage={onClickPage}
  />
);

Previous.propTypes = {
  previousPageExists: PropTypes.bool.isRequired,
  previousPage: PropTypes.number,
  onClickPage: PropTypes.func.isRequired,
};

Previous.defaultProps = {
  previousPage: null,
};

const Next = ({ nextPageExists, nextPage, onClickPage }) => (
  <Item
    id="next"
    page={nextPage}
    text="Next"
    isDisabled={!nextPageExists}
    onClickPage={onClickPage}
  />
);

Next.propTypes = {
  nextPageExists: PropTypes.bool.isRequired,
  nextPage: PropTypes.number,
  onClickPage: PropTypes.func.isRequired,
};

Next.defaultProps = {
  nextPage: null,
};


const Paginator = ({ onClickPage, paginator }) => {
  if (!paginator) return null;

  return (
    <div className="paginator">
      <Previous
        previousPageExists={paginator.previousPageExists}
        previousPage={paginator.previousPage}
        onClickPage={onClickPage}
      />
      <List
        pagesList={paginator.pagesList}
        currentPage={paginator.page}
        onClickPage={onClickPage}
      />
      <Next
        nextPageExists={paginator.nextPageExists}
        nextPage={paginator.nextPage}
        onClickPage={onClickPage}
      />
    </div>
  );
};

Paginator.propTypes = {
  onClickPage: PropTypes.func.isRequired,
  paginator: paginatorType,
};

Paginator.defaultProps = {
  paginator: null,
};

export default Paginator;
