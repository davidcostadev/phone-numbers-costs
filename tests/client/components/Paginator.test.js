import React from 'react';
import renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Paginator from '../../../client/components/Paginator';

configure({ adapter: new Adapter() });

describe('Paginator Component', () => {
  let onClickPage;
  let paginator;

  beforeEach(() => {
    onClickPage = jest.fn();
    paginator = {
      maxPage: 3,
      minPage: 1,
      nextPage: 2,
      nextPageExists: true,
      page: 1,
      pagesList: [1, 2, 3],
      perPage: 10,
      previousPage: null,
      previousPageExists: false,
      totalPages: 100,
    };
  });

  describe('Render', () => {
    it('default', () => {
      const tree = renderer.create((
        <Paginator onClickPage={onClickPage} />
      )).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('with props', () => {
      const tree = renderer.create((
        <Paginator onClickPage={onClickPage} paginator={paginator} />
      )).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('Behavior', () => {
    it('click on onClickPage', () => {
      const wrapper = mount((
        <Paginator onClickPage={onClickPage} paginator={paginator} />
      ));

      wrapper.find('#paginator-1').simulate('click');
      expect(onClickPage).toBeCalledWith(1);

      wrapper.find('#paginator-2').simulate('click');
      expect(onClickPage).toBeCalledWith(2);

      wrapper.find('#paginator-3').simulate('click');
      expect(onClickPage).toBeCalledWith(3);

      wrapper.find('#paginator-next').simulate('click');
      expect(onClickPage).toBeCalledWith(2);

      wrapper.find('#paginator-previous').simulate('click');
      expect(onClickPage).toBeCalledWith(2);
    });
  });
});
