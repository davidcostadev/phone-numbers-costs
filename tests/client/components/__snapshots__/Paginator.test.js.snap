// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator Component Render default 1`] = `null`;

exports[`Paginator Component Render with props 1`] = `
<div
  className="paginator"
>
  <button
    className="paginator__item paginator__item--disabled"
    disabled={true}
    id="paginator-previous"
    onClick={[Function]}
    type="button"
  >
    Previous
  </button>
  <button
    className="paginator__item paginator__item--active"
    disabled={false}
    id="paginator-1"
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className="paginator__item"
    disabled={false}
    id="paginator-2"
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className="paginator__item"
    disabled={false}
    id="paginator-3"
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    className="paginator__item"
    disabled={false}
    id="paginator-next"
    onClick={[Function]}
    type="button"
  >
    Next
  </button>
</div>
`;
